addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import babel.dates
+import logging
+import re
+
+from datetime import datetime, timedelta
+
+from odoo import _, models
+from odoo.exceptions import AccessDenied
+from odoo.http import request
+from odoo.tools.misc import babel_locale_parse
+
+from odoo.addons.auth_totp.models.totp import hotp, TOTP
+
+_logger = logging.getLogger(__name__)
+
+TOTP_RATE_LIMITS = {
+    'send_email': (10, 3600),
+    'code_check': (10, 3600),
+}
+
+
+class Users(models.Model):
+    _inherit = 'res.users'
+
+    def _mfa_url(self):
+        r = super()._mfa_url()
+        if r is not None:
+            return r
+        ICP = self.env['ir.config_parameter'].sudo()
+        otp_required = False
+        if ICP.get_param('auth_totp.policy') == 'all_required':
+            otp_required = True
+        elif ICP.get_param('auth_totp.policy') == 'employee_required' and self.env.user.has_groups('base.group_user'):
+            otp_required = True
+        if otp_required:
+            return '/web/login/totp'
+
+    def _totp_check(self, code):
+        self._totp_rate_limit('code_check')
+        user = self.sudo()
+        if user.totp_enabled:
+            return super()._totp_check(code)
+
+        key = self._get_totp_mail_key()
+        match = TOTP(key).match(code, window=3600, timestep=3600)
+        if match is None:
+            _logger.info("2FA check (mail): FAIL for %s %r", self, self.login)
+            raise AccessDenied(_("Verification failed, please double-check the 6-digit code"))
+        _logger.info("2FA check(mail): SUCCESS for %s %r", self, self.login)
+        self._totp_rate_limit_purge('code_check')
+        return True
+
+    def _get_totp_mail_key(self):
+        self.ensure_one()
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "Mail TOTP protection requires a configured database secret"
+        return f'{self.id}/{self.login}/{secret}'.encode()
+
+    def _get_totp_mail_code(self):
+        self.ensure_one()
+
+        key = self._get_totp_mail_key()
+
+        now = datetime.now()
+        counter = int(datetime.timestamp(now) / 3600)
+
+        code = hotp(key, counter)
+        expiration = timedelta(seconds=3600)
+        lang = babel_locale_parse(self.env.context.get('lang') or self.lang)
+        expiration = babel.dates.format_timedelta(expiration, lang)
+
+        return code, expiration
+
+    def _send_totp_mail_code(self):
+        self.ensure_one()
+        self._totp_rate_limit('send_email')
+
+        template = self.env.ref('auth_totp_mail_enforce.mail_template_totp_mail_code').sudo()
+        context = {}
+        if request:
+            geoip = request.session.geoip
+            context.update({
```

Comment:
```suggestion
            context = {
```
it's not like there's anything in the context before that.

Reply:
Well, the variable `context` needs to defined outside the `if` block as it's used outside it, in `template.with_context(**context).send_mail`
 and then choosing to re-define completely the variable or update it seems like a styling choice to me.
For instance, in case in the future we add something in the future by default in the context, leaving it the current way will lead to less diff.